src/contracts/pods/EigenPodManager.sol

```diff
@@ -206,7 +334,169 @@ contract EigenPodManager is Initializable, OwnableUpgradeable, Pausable, IEigenP
+        * By doing this, we ensure that the number of shares in EigenLayer matches the amount of withdrawable ETH in 
+        * the pod plus any ETH still staked on the beacon chain via other validators pointed to the pod. As a result, a validator 
+        * must complete a full withdrawal from the execution layer prior to queuing a withdrawal of 'beacon chain ETH shares' 
+        * via EigenLayer, since otherwise withdrawableRestakedExecutionLayerGwei will be 0.
+        */      
+        _decrementWithdrawableRestakedExecutionLayerGwei(podOwner, amountWei);  
+
+        address delegatedAddress = delegationManager.delegatedTo(podOwner);
+
+        uint96 nonce = uint96(numWithdrawalsQueued[podOwner]);
+        unchecked {
+            numWithdrawalsQueued[podOwner] = nonce + 1;
+        }
+
+        BeaconChainQueuedWithdrawal memory queuedWithdrawal = BeaconChainQueuedWithdrawal({
+            shares: amountWei,
+            podOwner: podOwner,
+            nonce: nonce,
+            withdrawalStartBlock: uint32(block.number),
+            delegatedAddress: delegatedAddress
+        });
+
+        // If the `staker` has withdrawn all of their funds from EigenLayer in this transaction, then they can choose to also undelegate
+        /**
+         * Checking that `stakerCanUndelegate` is not strictly necessary here, but prevents reverting very late in logic,
+         * in the case that `undelegateIfPossible` is set to true but the `staker` still has active deposits in EigenLayer.
+         */
+        if (undelegateIfPossible && stakerCanUndelegate(podOwner)) {
+            _undelegate(podOwner);
+        }
+
+        bytes32 withdrawalRoot = calculateWithdrawalRoot(queuedWithdrawal);
+        withdrawalRootPending[withdrawalRoot] = true;
+
+        emit BeaconChainETHWithdrawalQueued(podOwner, amountWei, nonce);   
+
+        return withdrawalRoot;
+    }
+
+    // TODO: add documentation to this function
+    function _completeQueuedWithdrawal(
+        BeaconChainQueuedWithdrawal memory queuedWithdrawal,
+        uint256 middlewareTimesIndex
+    )
+        internal
+    {
+        // find the withdrawalRoot
+        bytes32 withdrawalRoot = calculateWithdrawalRoot(queuedWithdrawal);
+
+        // verify that the queued withdrawal is pending
+        require(
+            withdrawalRootPending[withdrawalRoot],
+            "EigenPodManager._completeQueuedWithdrawal: withdrawal is not pending"
+        );
+
+        // verify that the withdrawal is completable
+        require(
+            slasher.canWithdraw(queuedWithdrawal.delegatedAddress, queuedWithdrawal.withdrawalStartBlock, middlewareTimesIndex),
+            "EigenPodManager._completeQueuedWithdrawal: shares pending withdrawal are still slashable"
+        );
+
+        /* TODO: decide definitively if minimum lag is enforced here or via use of the DelayedWithdrawalRouter
```

Comment:
here

Reply:
> here

I agree this feels more "correct" to me. please review this commit and comment yea or nay on it https://github.com/Layr-Labs/eigenlayer-contracts/pull/135/commits/7fa28e892f9e86190e297fe81205dbca25526066